File: src_python/network_conf_server.py
import os
import argparse
import logging
import time
from pathlib import Path
from flask import request
from configparser import ConfigParser
from flask import Flask, render_template, jsonify

from interface_manager.inteface_manager import InterfaceManager

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
logger_werkzeug = logging.getLogger("werkzeug")
logger_werkzeug.setLevel(logging.INFO)

VERSION = "1.0"

USE_FULL_BACKTRACE = True

DEFAULT_CONFIG = "network-configuration.default.conf"

DESCRIPTION = f"""

"""

root_dir = Path(".")

class ReverseProxied(object):

    def __init__(self, app, script_name=None, scheme=None, server=None):
        self.app = app
        self.script_name = script_name
        self.scheme = scheme
        self.server = server

    def __call__(self, environ, start_response):
        script_name = environ.get('HTTP_X_SCRIPT_NAME', '') or self.script_name
        if script_name:
            environ['SCRIPT_NAME'] = script_name
            path_info = environ['PATH_INFO']
            if path_info.startswith(script_name):
                environ['PATH_INFO'] = path_info[len(script_name):]
        scheme = environ.get('HTTP_X_SCHEME', '') or self.scheme
        if scheme:
            environ['wsgi.url_scheme'] = scheme
        server = environ.get('HTTP_X_FORWARDED_SERVER', '') or self.server
        if server:
            environ['HTTP_HOST'] = server
        return self.app(environ, start_response)


class NetworkConfigurationService:
    def __init__(self, def_config):
        self.manager = InterfaceManager(def_config=def_config)
        self._start_server = def_config.getboolean('Server', 'EnableServer')
        self._port = def_config.getint('Server', 'Port')
        self._address = def_config.get('Server', 'Address')
        self._ap_hide_in_ui = def_config.getboolean('AP', 'APHideInUI')
        self._ap_interface = def_config.get('AP', 'APInterfaceDevice')
        self._reverse_proxy_path = def_config.get('Server', 'ReverseProxyPath')
        if self._start_server:
            self.start_server()
        else:
            # Just sleep indefinetely
            while True:
                time.sleep(100)

    def start_server(self):
        app = Flask(__name__,
                    static_folder=root_dir / 'static')
        app.wsgi_app = ReverseProxied(app.wsgi_app, script_name=self._reverse_proxy_path)

        @app.route('/')
        def index():
            return app.send_static_file('index.html')

        @app.route('/api/status', methods=['GET'])
        def status_control():
            return jsonify(self.manager.get_status()), 200

        @app.route('/api/config', methods=['GET', 'POST'])
        def config_control():
            if request.method == 'GET':
                conf = self.manager.get_conf()
                if self._ap_hide_in_ui:
                    try:
                        conf.pop(self._ap_interface)
                    except:
                        pass
                return jsonify(conf), 200
            elif request.method == 'POST':
                config = request.get_json()
                logger.info(f"Received config: {config}")
                self.manager.load_config(config)
                return jsonify("OK"), 200

        @app.route('/api/<interface_id>/config', methods=['GET', 'POST'])
        def config_interface_control(interface_id):
            try:
                for interface in self.manager.interfaces:
                    if interface_id == interface.device:
                        if request.method == 'GET':
                            return jsonify(interface.get_conf()), 200
                        elif request.method == 'POST':
                            config = request.get_json()
                            logger.info(f"Received config: {config}")
                            interface.load_config({interface_id: config})
                            return jsonify("OK"), 200
                return jsonify({'error': f'Interface {interface_id} not found. Acceptable interfaces are: {", ".join(self.manager.interfaces)}'}), 404
            except Exception as e:
                return jsonify({'error': f'{e}'}), 500

        @app.route('/api/interfaces', methods=['GET'])
        def interfaces_control():
            interfaces = []
            for interface in self.manager.interfaces:
                if self._ap_hide_in_ui and interface.device == self._ap_interface:
                    continue
                interfaces.append(interface.device)
            return jsonify(interfaces), 200

        @app.route('/api/param/<interface_id>/<parameter>', methods=['GET', 'POST'])
        def parameter_control(interface_id: str, parameter: str):
            try:
                for interface in self.manager.interfaces:
                    if interface_id == interface.device:
                        if parameter in interface.parameters():
                            if request.method == 'GET':
                                return jsonify(interface[parameter]), 200
                            elif request.method == 'POST':
                                try:
                                    interface[parameter] = request.form.to_dict(flat=False)
                                    return jsonify(interface[parameter]), 200
                                except Exception as e:
                                    return jsonify({'error': f'Could not process request, internal error: {e}'}), 500
                            else:
                                return jsonify({'error': f'Method {request.method} not allowed'}), 405
                        else:
                            return jsonify({'error': f'Unknown parameter {parameter} for interface {interface_id}. '
                                                     f'Acceptable parameters are: {", ".join(interface.parameters())}'}), 404
                    return jsonify({'error': f'Interface {interface_id} not found. Acceptable interfaces are: {", ".join(self.manager.interfaces)}'}), 404
            except Exception as e:
                return jsonify({'error': f'{e}'}), 500

        app.run(debug=False, port=self._port, host=self._address)

    def run(self):
        self.start_server()


def main():
    parser = argparse.ArgumentParser(prog=f"Network Configuration Service {VERSION}",
                                     description=DESCRIPTION)
    parser.add_argument("-c", "--conf", help="Configuration file", type=str, required=False)
    args = parser.parse_args()

    try:
        def_config = ConfigParser()
        def_config.read(DEFAULT_CONFIG)
        if os.path.isfile(args.conf):
            def_config.read(args.conf)
        else:
            logger.warning("Configuration file not provided, using default configuration")

        NetworkConfigurationService(def_config)
    except Exception as e:
        logger.error(f"Exception: {e}")
        if USE_FULL_BACKTRACE:
            raise e


if __name__ == "__main__":
    main()

File: src_go/network_conf_server.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/bigkevmcd/go-configparser"
)

const VERSION = "1.0"
const DEFAULT_CONFIG_FILE = "network-configuration.default.conf"
const STATIC_CONTENT_PATH = "static"

type NetworkConfigurationService struct {
	manager           *InterfaceManager
	port              int
	address           string
	apHideInUI        bool
	apInterface       string
	reverseProxyPath  string
	startServer       bool
}

func (n *NetworkConfigurationService) StartServer() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		http.ServeFile(w, r, fmt.Sprintf("%s/index.html", STATIC_CONTENT_PATH))
	})

	http.HandleFunc("/api/status", func(w http.ResponseWriter, r *http.Request) {
		json.NewEncoder(w).Encode(n.manager.GetStatus())
	})

	http.HandleFunc("/api/config", func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodGet {
			conf := n.manager.GetConf()
			if n.apHideInUI {
				delete(conf, n.apInterface)
			}
			json.NewEncoder(w).Encode(conf)
		} else if r.Method == http.MethodPost {
			var config map[string]interface{}
			if err := json.NewDecoder(r.Body).Decode(&config); err != nil {
				http.Error(w, err.Error(), http.StatusBadRequest)
				return
			}
			log.Printf("Received config: %v", config)
			n.manager.LoadConfig(config)
			w.Write([]byte("OK"))
		}
	})

	http.HandleFunc("/api/interfaces", func(w http.ResponseWriter, r *http.Request) {
		var interfaces []string
		for _, iface := range n.manager.Interfaces {
			if n.apHideInUI && iface.Device == n.apInterface {
				continue
			}
			interfaces = append(interfaces, iface.Device)
		}
		json.NewEncoder(w).Encode(interfaces)
	})

	address := fmt.Sprintf("%s:%d", n.address, n.port)
	log.Printf("Starting server on %s", address)
	if err := http.ListenAndServe(address, nil); err != nil {
		log.Fatalf("Server failed: %v", err)
	}
}

func main() {
	var configurationFile = flag.String("c", DEFAULT_CONFIG_FILE, "Configuration file")
	flag.Parse()

	p, err := configparser.NewConfigParserFromFile(configurationFile)
	if err != nil {
		log.Fatalf("Failed to load config (%v): %v", configurationFile, err)
	}

	port, _ := p.GetInt("Server", "Port")
	address, _ := p.Get("Server", "Address")
	startServer, _ := p.GetBool("Server", "EnableServer")
	apHideInUI, _ := p.GetBool("AP", "APHideInUI")
	apInterface := p.Get("AP", "APInterfaceDevice")
	reverseProxyPath := p.Get("Server", "ReverseProxyPath")

	service := NetworkConfigurationService{
		manager: &InterfaceManager{},
		port:              port,
		address:           address,
		apHideInUI:        apHideInUI,
		apInterface:       apInterface,
		reverseProxyPath:  reverseProxyPath,
		startServer:       startServer,
	}

	if startServer {
		service.StartServer()
	} else {
		log.Println("Server disabled, sleeping indefinitely...")
		for {
			time.Sleep(100 * time.Second)
		}
	}
}
